### PR TITLE
Added Snyk Code Parser

### PR DESCRIPTION
**Pull request by FelixHernandez**
##  Added Snyk Code Parser
[sc-4509]

**Description**

A new `SnykCodeParser` was added and the deduplication is based on `ruleID` and `URI` (`file_path`).
The unittests for this parser were added as well.
https://github.com/DefectDojo/django-DefectDojo/issues/9604

### CODE REVIEW

File: docs/content/en/integrations/parsers/file/snyk_code.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+---
+title: "Snyk"
```
Comment (manuel-sommer) [inline docs/content/en/integrations/parsers/file/snyk_code.md:2]: Hi @FelixHernandez 
title should then be "Snyk Code"

In general I personally like it more if we don't add more and more parsers for multiple functionalities of one vendor, but maybe rather one parser which covers all functionalities of a vendor. This is why I made PRs like these two:
https://github.com/DefectDojo/django-DefectDojo/pull/9522
https://github.com/DefectDojo/django-DefectDojo/pull/9355

Furthermore: Did you check out this PR:
https://github.com/DefectDojo/django-DefectDojo/pull/9289

File: dojo/tools/snyk_code/parser.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
+import json
+
+from cvss.cvss3 import CVSS3
+from dojo.models import Finding
+
+
+class SnykCodeParser(object):
+    def get_scan_types(self):
+        return ["Snyk Code Scan"]
```
Comment (manuel-sommer) [inline dojo/tools/snyk_code/parser.py:9]: if you implemented this parser to distinguish between Snyk and Snyk Code, I would rather recommend to either do it like it was done with RustyHog: https://github.com/DefectDojo/django-DefectDojo/blob/master/dojo/tools/rusty_hog/parser.py or with Checkmarx: https://github.com/DefectDojo/django-DefectDojo/blob/c5d896b2a606985f41669388962e47409f4866ba/dojo/tools/checkmarx/parser.py#L16

**Review by manuel-sommer — CHANGES_REQUESTED** [review]

**Comment by manuel-sommer** [comment]
Hi @grendel513,

I would like to understand why you resolved my review. Could you please comment on that?

Best Regards

**Comment by grendel513** [comment]
> Hi @grendel513,
> 
> I would like to understand why you resolved my review. Could you please comment on that?
> 
> Best Regards

My comment specifically stated why: Keeping this a separate parser is the correct implementation to allow for its own deduplication which is what we are trying to solve for. Additionally having to individual parsers is easier to troubleshoot should issues arise, and has a the benefit of not affecting other tool ingestion if there is something wrong.

Moving forward we will be asking that new parsers be single threaded in function, as having parsers that parse multiple file types creates issues with debugging, deduplication, and the ability to have a separate deduplication configuration.

**Comment by manuel-sommer** [comment]
> > Hi @grendel513,
> > I would like to understand why you resolved my review. Could you please comment on that?
> > Best Regards
> 
> My comment specifically stated why: Keeping this a separate parser is the correct implementation to allow for its own deduplication which is what we are trying to solve for. Additionally having to individual parsers is easier to troubleshoot should issues arise, and has a the benefit of not affecting other tool ingestion if there is something wrong.
> 
> Moving forward we will be asking that new parsers be single threaded in function, as having parsers that parse multiple file types creates issues with debugging, deduplication, and the ability to have a separate deduplication configuration.

Thank you for clarification. 😄 I understand your logic. 
What if we advance the parser structure in a way that file stuctures are automatically recognized (e.g. look at RustyHog) and then deduplication could be implemented for the subfile types (e.g. Deduplication for GottingenHog (RustyHog) etc.)
This whole topic is pretty complicated as there are pros and cons in both directions, but for beginners it would be easier if the number of parsers to be selected within a vendor is limited and the logic within DefectDojo handles the rest. 
One way could also be which was implemented in veracode:
1. entry is a single parser for a vendor
2. parser recognizes the file type (subscanner)
3. individual deduplication and scannertype recognition (e.g. like in veracode + rustyhog)

**Review by mtesauro — APPROVED** [review]
Approved

**Review by grendel513 — APPROVED** [review]

**Review by blakeaowens — APPROVED** [review]

**Review by cneill — APPROVED** [review]

**Pull request merged**